java/maven/src/org/netbeans/modules/maven/options/MavenSettings.java

```diff
@@ -480,6 +481,18 @@ public void setBinaryDownloadStrategy(DownloadStrategy ds) {
             }
         }
 
+        try {
+            try (Stream<Path> jars = Files.list(lib).filter(file -> file.getFileName().toString().startsWith("maven-core"))) {  // NOI18N
+                for (Path mavenCoreJar : jars.collect(Collectors.toList())) {
+                    String mavenCoreJarName = mavenCoreJar.getFileName().toString();
+                    if (mavenCoreJarName.matches("maven-core-\\d+\\.\\d+\\.\\d+\\.jar")) {
+                        return mavenCoreJarName.replace("maven-core-", "").replace(".jar", "");
+                    }
+                }
+            }
+        } catch (IOException ignored) {
+        }
```

Comment:
just a nitpick: it would be even faster by terminating the stream early (and not collecting the jars which forces the stream to visit all jars):
```java
        Pattern jarPattern = Pattern.compile("maven-core-\\d+\\.\\d+\\.\\d+\\.jar");  // NOI18N
        try (Stream<Path> jars = Files.list(lib)
                                      .map(file -> file.getFileName())
                                      .filter(file -> file.toString().startsWith("maven-core"))  // NOI18N
                                      .filter(file -> jarPattern.matcher(file.toString()).matches())) {
            Optional<Path> mavencore = jars.findFirst();
            if (mavencore.isPresent()) {
                return mavencore.get().toString().replace("maven-core-", "").replace(".jar", "");  // NOI18N
            }
        } catch (IOException ignored) {}
```
this uses one stream start to finish and it should terminate as soon the jar is found. The pattern is now also compiled outside of the loop.

I did also attach the catch to the inner try so that we only need one try block for this.

(isn't tested, hope it compiles)

Reply:
You're definitely right. I guess I took too much from the previous loop/check.

I took your suggestion and modified it a little bit further. I thought it might be better to reuse the `Pattern` if we go this route so I changed to code to extract the version from that `Pattern` as well.

Maybe you can have a second look at the code and open a new suggestion on that part if necessary.

Reply:
i like it, thanks Benjamin!